test/test_utilities.go

```diff
@@ -0,0 +1,92 @@
+package test
+
+import (
+	"fmt"
+	"github.com/Shopify/sarama"
+	"log"
+	"regexp"
+	"sync"
+	"time"
+)
+
+
+type ExampleConsumerGroupHandler struct{
+	// mutex for exclusive write on message handler
+	mutexWritePartitionPresence *sync.Mutex
+	consumingDone               chan bool
+	partitionsConsumptionSlice  []bool
+
+}
+
+func NewConsumerGroupHandler() ExampleConsumerGroupHandler {
+	handler := ExampleConsumerGroupHandler{}
+	handler.mutexWritePartitionPresence = &sync.Mutex{}
+	handler.consumingDone = make(chan bool)
+	return handler
+}
+
+func (h ExampleConsumerGroupHandler) isEveryPartitionConsumed() bool  {
+	for _,value := range h.partitionsConsumptionSlice {
+		if value != true {
+			return false
+		}
+	}
+	return true
+}
+
+func (ExampleConsumerGroupHandler) Setup(_ sarama.ConsumerGroupSession) error { return nil }
+
+func (ExampleConsumerGroupHandler) Cleanup(_ sarama.ConsumerGroupSession) error { return nil }
+
+func (h ExampleConsumerGroupHandler) ConsumeClaim(sess sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	for msg := range claim.Messages() {
+		h.mutexWritePartitionPresence.Lock()
+		// setting for current partition that at least one message has been produced.
+		h.partitionsConsumptionSlice[msg.Partition] = true
+		// Check that Consumer group consumed at least one message for each partition on canary topic
+		log.Println("msg. arrived")
+		if h.isEveryPartitionConsumed() {
+			log.Println("all partitions consumed")
+			h.consumingDone <- true
+		}
+		h.mutexWritePartitionPresence.Unlock()
+		sess.MarkMessage(msg, "")
+
+	}
+	return nil
+}
+
+// We are only interested in counter that is produced
+func parseSucReqRateFromMetrics( input string) string  {
+	regex, _ := regexp.Compile("(?m)^promhttp_metric_handler_requests_total.*200...(\\d+)$")
+	data := regex.FindStringSubmatch(input)
+	if len(data) > 1 {
+		return data[1]
+	}
+	return ""
+}
+
+// get total number of produced records for canary test topic
+func parseCanaryRecordsProducedFromMetrics( input string) string {
+	regex, _ := regexp.Compile("(?m)^.*strimzi_canary_records_produced_total\\S*\\s(\\d+)$")
+	data := regex.FindStringSubmatch(input)
+	if len(data) > 1 {
+		return data[1]
+	}
+	return ""
+}
+
+func getPartitionCount( clusterAdmin sarama.ClusterAdmin ,topicName string ) int{
+	for i := 0; i < 20; i++ {
```

Comment:
why this loop? ... also, you have already called `DescribeTopics` in the function test which calls this function `getPartitionCount` so why doing it again? The topic.go file shows that a describe topics is enough to get the topic and the partitions number.

Reply:
because this value needs to be updated, properly. asking too soon ends up in answering with 0 value.

Reply:
I assume that it is due to asynchronous nature of Kafka, and probably I am asking too soon since topic is created ?

Reply:
after second or so provided response match expectation but usually on first attempt empty list is provided.

Reply:
but in the first DescribeTopics, the topic exists otherwise your test fails correctly ... so it should already have the number of partitions, it doesn't make sense that the admin call reports back metadata for a topic but no info about partitions.

Reply:
![image](https://user-images.githubusercontent.com/40673682/128878152-88326b9e-6fb7-4da2-a1a5-136222de1b7b.png)

Maybe at least single data produced in that partition I honestly don't know. As I wrote few days later, using old approach it returned correct number immediately.

Reply:
I will find the time to try/run this and check what Sarama is doing underneath, unless you already debugged it.

Reply:
yes I see, I can retry directly here instead of later with second Admin :)

Reply:
your tests cannot start until the topic is there (no Err) and you know the number of partitions so it would be the first thing to do with just one admin client at the beginning.